src/core/ruby_version.rs

```diff
@@ -919,6 +1003,107 @@ macro_rules! get_stack_frame_2_5_0(
     )
 );
 
+macro_rules! get_stack_frame_3_3_0(
+    () => (
+        fn get_classname<T>(
+            ep: &*const usize,
+            source: &T,
+        ) -> Result<String> where T: ProcessMemory {
+            let mut ep = ep.clone() as *mut usize;
+
+            let frame_flag: usize = unsafe {
+                source.copy_struct(ep.offset(0) as usize).context(ep.offset(0) as usize)?
+            };
+
+            // VM_FRAME_MAGIC_MASK   = 0x7fff0001
+            // VM_FRAME_MAGIC_METHOD = 0x11110001
+            if (frame_flag & 0x7fff0001) != 0x11110001 {
+                return Err(anyhow::anyhow!(format!("frame flag {:x} does not match method mask, cannot read class", frame_flag)));
+            }
+
+            let mut env_specval: usize = unsafe {
+                source.copy_struct(ep.offset(-1) as usize).context(ep.offset(-1) as usize)?
+            };
+            let mut env_me_cref: usize = unsafe {
+                source.copy_struct(ep.offset(-2) as usize).context(ep.offset(-1) as usize)?
+            };
+
+            // #define VM_ENV_FLAG_LOCAL 0x02
+            while env_specval & 0x02 != 0 {
+                if !check_method_entry(env_me_cref, source)?.is_null() {
+                    break;
+                }
+                unsafe {
+                    ep = env_specval.clone() as *mut usize;
+                    env_specval = source.copy_struct(ep.offset(-1) as usize).context(ep.offset(-1) as usize)?;
+                    env_me_cref = source.copy_struct(ep.offset(-2) as usize).context(ep.offset(-2) as usize)?;
+                }
+            }
+
+            let class_name = match source.copy_struct(env_me_cref).context(env_me_cref) {
```

Comment:
This flattened version leans on the `?` operator and seems to work:

```rust
let imemo: rb_method_entry_struct = source.copy_struct(env_me_cref).context(env_me_cref)?;
let klass: RClass_and_rb_classext_t = source.copy_struct(imemo.defined_class).context(imemo.defined_class)?;
if klass.classext.classpath == 0 {
    return Err(anyhow::anyhow!("classpath was empty"));
}
let class_name = get_ruby_string(klass.classext.classpath as usize, source)?;
Ok(class_name.to_string())
```

Reply:
@acj thanks i'll apply it and give it a go